Fix bug 1063165 - add authentication

jwhitlock opened the pull request:
* Add django-allauth for social login framework
* Add Firefox Account relier
* Style base template for anon / logged-in state
* Style sample allauth templates
* Add autoenv sample config
* Fix bug 1133066 - Better documentation, referencing autoenv sample config

groovecoder commented:
How do I set up the Firefox Accounts social application for django-allauth?

jwhitlock commented:
It's super simple:

1. Become a Mozilla employee
2. Provision developer FxA credentials with the [FxA OAuth Dashboard](https://developer.mozilla.org/en-US/Firefox_Accounts_OAuth_Dashboard).  Endpoint will be something like `http://localhost:8000/accounts/fxa/login/callback/`
3. Follow [allauth post-install instructions](http://django-allauth.readthedocs.org/en/latest/installation.html#post-installation) to add FxA credentials to database
4. (Optional) Setup an email server.  I went with [Amazon SES](https://aws.amazon.com/ses/) for the server.
5. Customize environment variables for the mail server and the [developer FxA endpoints](https://developer.mozilla.org/en-US/Firefox_Accounts#Firefox_Accounts_deployments), such as:
```
export FXA_OAUTH_ENDPOINT=https://oauth-latest.dev.lcip.org/v1
export FXA_PROFILE_ENDPOINT=https://latest.dev.lcip.org/profile/v1
export EMAIL_BACKEND=django.core.mail.backends.console.EmailBackend
```
Restart your server with the new environment variables, and you're done!

groovecoder commented:
Thanks. We should add (some of that) to docs

For now I'm stuck on https://github.com/mozilla/fxa-oauth-console/issues/23.

jwhitlock commented:
I'll add a docs commit to the PR in a few hours

jwhitlock commented:
Updated "Improve installation docs" commit w/ Firefox Accounts instructions

groovecoder commented:
I'm stuck on https://github.com/mozilla/fxa-oauth-console/issues/25 trying to spot-check this on my local box. (How) did you get past this with your local box?

groovecoder commented:
This is my .env file:
```
export DATABASE_URL="postgresql://lcrouch@localhost:5432/wpc_test"
export DJANGO_DEBUG=1
export PGDATA="/usr/local/var/postgres"
export FXA_OAUTH_ENDPOINT=https://oauth-stable.dev.lcip.org/v1
export FXA_PROFILE_ENDPOINT=https://stable.dev.lcip.org/profile/v1
export EMAIL_BACKEND=django.core.mail.backends.console.EmailBackend
```
And I'm still getting the behavior in https://github.com/mozilla/fxa-oauth-console/issues/25

groovecoder commented:
Argh! I updated `.env` but I didn't cd out of the directory and back in. One of the reasons I like [`python-decouple`](https://pypi.python.org/pypi/python-decouple) for environment variables now.

groovecoder commented on bcauth/helpers.py:
```diff
@@ -0,0 +1,58 @@
+
+    Jingo version of providers_media_js from
+    allauth/socialaccount/templatetags/socialaccount.py
+    """
```
Annoying that we have these jingo helper versions duplicated here and [in the kuma codebase](https://github.com/mozilla/kuma/blob/master/kuma/users/helpers.py#L83-L110). Seems like there could be a django-allauth-jingo project to help users of both.

jwhitlock replied:
I imagine jingo will be deprecated with Django 1.8: https://docs.djangoproject.com/en/dev/topics/templates/.  I'd focus kuma efforts on getting to a current Django version.

groovecoder commented on requirements.txt:
```diff
@@ -76,11 +76,15 @@ django-sortedm2m==0.8.1
 parsimonious==0.6.2
+
+# Social loging support (Firefox Accounts)
+python-openid==2.2.5
```
Really? We need openid for Firefox Accounts? Are these underlying libraries not included in the `django-allauth` requirement?

jezdez replied:
You can skip that requirement if you're not using the openid providers in allauth.

jwhitlock replied:
django-allauth [requires python-openid](https://github.com/pennersr/django-allauth/blob/master/setup.py#L124-L127) at install time.  Yes, it goes unused if you don't use open ID, but it does get installed.  My process is to add pinned versions of all installed requirements.  I think it is to minimize the surprise when a new dev configures their environment a few months after I do, and gets different packages with different behaviour.  This may be good practice or a voodoo luck charm.  What does kuma do?

I'll occasionally run something like this to check packages:

```sh
pip freeze > installed.txt
cat requirements* | sort | grep -v "^$" | grep -v "^#" > intended.txt
diff installed.txt intended.txt
```

As for the requirements file, I list the requirements in install order - `oauthlib` and `requests` are installed as prerequisites for `requests-oauthlib`, which is needed for `django-allauth`.  I group them with comments, a short explanation of why the group of requirements is there.

jwhitlock replied:
And looking a little further up, there's a [different requirement](https://github.com/pennersr/django-allauth/blob/master/setup.py#L103-L107) for py3.  I don't think pip has a way to mark a requirement as a py2 vs py3 requirement, so I'll probably comment out the line.

groovecoder replied:
Kuma isn't the best example since we use submodules. But that means we pin all dependencies at specific commits/versions. It avoids new-user/developer surprise, but it means dependencies fall out of date more easily.

In any case, since `django-allauth` lists it in their requirements, it makes sense to leave it here for consistency. So I rescind my original comment. :smile:

groovecoder commented:
I'm ready to merge if we include the "Choose a site" step in the docs.

jwhitlock commented:
I'm going to move the Firefox Account docs to the wiki.  It's hard to setup, not needed for development, and only @mozilla.com employees can actually setup credentials.

groovecoder commented:
Makes sense. Link to the wiki from the docs site?

jwhitlock commented:
`git push -f`:

* Commented out `python-openid` in requirements.txt - let django-allauth pick depending on python version
* Added wiki link to README / docs
* Moved the FxA instructions to [the wiki](https://github.com/jwhitlock/web-platform-compat/wiki/Firefox%20Accounts)

One more change coming - fetching `/accounts/profile` as an anon user is a 500.
